### PR TITLE
Introduce asset filter for Category grid, Page Blocks and product grid

### LINKED ISSUE #1479: Add media asset filter to the product grid
**Story** https://github.com/magento/adobe-stock-integration/issues/1292

### Preconditions (*)
1. Install Magento 2.4-develop

### Steps to reproduce (*)
1. Go to Catalog -> Products
2. Expand Filters section

### Expected result (*)
Media Asset filter is present. The filter can be used to search the products where specific media asset is used

image

image

### Technical details

The filters should be added to MediaContentCatalog module in magento2 repo
The products for which the asset is used can be identified using the `\Magento\MediaContentApi\Api\GetContentByAssetIdsInterface` service

### AC
- covered with MFTF test

### LINKED ISSUE #1480: Add Media Asset filter to Cms page and block grids
**Story** https://github.com/magento/adobe-stock-integration/issues/1292

### Preconditions (*)
1. Install Magento 2.4-develop

### Steps to reproduce (*)
1. Go to Cms -> Pages (Cms -> Blocks)
2. Expand Filters section

### Expected result (*)
Media Asset filter is present. The filter can be used to search the products where specific media asset is used

image

image

### Technical details

The filters should be added to MediaContentCms module in magento2 repo
The blocks/pages for which the asset is used can be identified using the `\Magento\MediaContentApi\Api\GetContentByAssetIdsInterface` service

### AC
- covered with MFTF test

### LINKED ISSUE #1502: Add asset filter to category grid
Asset filter should be present for category grid on `admin/media_gallery_catalog/category/index` page

Depends on https://github.com/magento/adobe-stock-integration/issues/1481

### PR DESCRIPTION
**Pull request by Nazar65**
### Fixed Issues (if relevant)

1. Fixes magento/adobe-stock-integration#1502: add asset filter to category grid
2. Fixes magento/adobe-stock-integration#1479 Add media asset filter to the product grid 
3. Fixes magento/adobe-stock-integration#1480 Add Media Asset filter to Cms page and block grids

### Manual testing scenarios (*)
1.Open Category grid page by link `admin/media_gallery_catalog/category/index`
2. Expand Filters tab
3. Put asset path in Asset filter input
4. Same for product grid , pages grid , blocks grid

Screenshot_2020-07-10 Categories Magento Admin

### CODE REVIEW

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

File: MediaGalleryUi/view/adminhtml/web/js/form/element/ui-select.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    'use strict';
+
+    return uiSelect.extend({
+        /**
+         * Get path to current option
+         *
+         * @param {Object} data - option data
+         * @returns {String} path
+         */
+        getPath: function (data) {
```
Comment (sivaschenko) [inline MediaGalleryUi/view/adminhtml/web/js/form/element/ui-select.js]: Can we avoid the need to extend the component by using showPath=false configuration and just using a custom template

File: MediaGalleryUi/view/adminhtml/ui_component/product_listing.xml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+                component="Magento_MediaGalleryUi/js/form/element/ui-select"
+                template="Magento_MediaGalleryUi/grid/filters/elements/ui-select">
+                <argument name="data" xsi:type="array">
+                    <item name="config" xsi:type="array">
+                        <item name="entityType" xsi:type="string">catalog_product</item>
+                        <item name="identityColumn" xsi:type="string">entity_id</item>
+                        <item name="chipsEnabled" xsi:type="boolean">true</item>
+                        <item name="filterOptions" xsi:type="boolean">true</item>
+                        <item name="searchOptions" xsi:type="boolean">true</item>
+                        <item name="searchUrl" xsi:type="url" path="media_gallery/image/getassetsoptions" />
```
Comment (sivaschenko) [inline MediaGalleryUi/view/adminhtml/ui_component/product_listing.xml]: Let's use the following naming for consistency
```suggestion
                        <item name="searchUrl" xsi:type="url" path="media_gallery/asset/search" />
```

File: MediaGalleryUi/view/adminhtml/ui_component/product_listing.xml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+-->
+<listing xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:module:Magento_Ui:etc/ui_configuration.xsd">
+    <listingToolbar name="listing_top">
+        <filters name="listing_filters">
+              <filterSelect
+                name="asset"
```
Comment (sivaschenko) [inline MediaGalleryUi/view/adminhtml/ui_component/product_listing.xml]:
```suggestion
                name="asset_id"
```

File: MediaGalleryUi/view/adminhtml/ui_component/product_listing.xml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+                provider="${ $.parentName }"
+                sortOrder="10"
+                class="Magento\MediaGalleryUi\Ui\Component\Listing\Filters\Asset"
+                component="Magento_MediaGalleryUi/js/form/element/ui-select"
+                template="Magento_MediaGalleryUi/grid/filters/elements/ui-select">
+                <argument name="data" xsi:type="array">
+                    <item name="config" xsi:type="array">
+                        <item name="entityType" xsi:type="string">catalog_product</item>
+                        <item name="identityColumn" xsi:type="string">entity_id</item>
+                        <item name="chipsEnabled" xsi:type="boolean">true</item>
```
Comment (sivaschenko) [inline MediaGalleryUi/view/adminhtml/ui_component/product_listing.xml]: Please add 

                        <item name="filterPlaceholder" xsi:type="string" translate="true">Asset Title</item>
                        <item name="emptyOptionsHtml" xsi:type="string" translate="true">Start typing to find assets</item>

File: MediaGalleryUi/Test/Mftf/ActionGroup/AdminMediaGalleryAssertCategoryNameInCategoryGridActionGroup.xml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+              xsi:noNamespaceSchemaLocation="urn:magento:mftf:Test/etc/actionGroupSchema.xsd">
+    <actionGroup name="AdminMediaGalleryAssertCategoryNameInCategoryGridActionGroup">
+        <annotations>
+            <description>Asserts category name in category grid page</description>
+        </annotations>
+        <arguments>
+            <argument name="categoryName" type="string"/>
+        </arguments>
+
+
```
Comment (sivaschenko) [inline MediaGalleryUi/Test/Mftf/ActionGroup/AdminMediaGalleryAssertCategoryNameInCategoryGridActionGroup.xml]: unnecessary empty line

File: MediaGalleryUi/Controller/Adminhtml/Image/GetAssetsOptions.php [hunk 6]
```diff
@@ -0,0 +1,171 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\MediaGalleryUi\Controller\Adminhtml\Image;
```
Comment (sivaschenko) [inline MediaGalleryUi/Controller/Adminhtml/Image/GetAssetsOptions.php]:
```suggestion
namespace Magento\MediaGalleryUi\Controller\Adminhtml\Asset;
```

File: MediaGalleryUi/Controller/Adminhtml/Image/GetAssetsOptions.php [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+use Magento\Framework\Api\SearchCriteriaBuilder;
+use Magento\Framework\Api\FilterBuilder;
+use Magento\Cms\Helper\Wysiwyg\Images;
+use Magento\Cms\Model\Wysiwyg\Images\Storage;
+use Magento\Framework\Api\Search\FilterGroupBuilder;
+
+/**
+ * Controller getting the asset options for multiselect filter
+ */
+class GetAssetsOptions extends Action implements HttpGetActionInterface
```
Comment (sivaschenko) [inline MediaGalleryUi/Controller/Adminhtml/Image/GetAssetsOptions.php]:
```suggestion
class Search extends Action implements HttpGetActionInterface
```

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Thanks for the PR @Nazar65 ! Please see my review comments

File: MediaGalleryUi/Ui/Component/Listing/Filters/Asset.php [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+use Magento\Ui\Component\Filters\FilterModifier;
+use Magento\MediaContentApi\Api\GetContentByAssetIdsInterface;
+use Magento\Framework\Data\OptionSourceInterface;
+
+/**
+ * Asset  filter
+ */
+class Asset extends Select
+{
+    private const CTAEGORY_ENTITY_TYPE = 'catalog_category';
```
Comment (gabrieldagama) [inline MediaGalleryUi/Ui/Component/Listing/Filters/Asset.php]: We can probably remove this constant.

```suggestion
```

File: MediaGalleryUi/Ui/Component/Listing/Filters/Asset.php [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+            $this->getContext()->getDataProvider()->addFilter($filter);
+        }
+    }
+
+    /**
+     * Return category ids by asset path.
+     *
+     * @param string $ids
+     */
+    private function getCategoryIdsByAsset(array $ids): string
```
Comment (gabrieldagama) [inline MediaGalleryUi/Ui/Component/Listing/Filters/Asset.php]: I guess this wasn't meant to be only for categories right? We might want to rename this one.
```suggestion
    private function getEntityIdsByAsset(array $ids): string
```

**Comment by Nazar65** [comment]
@magento run all tests

**Review by gabrieldagama — CHANGES_REQUESTED** [review]
Hi @Nazar65, thanks for the PR, it looks great! Please follow some small comments.

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Comment by Nazar65** [comment]
@magento run all tests

**Review by sivaschenko — APPROVED** [review]

**Comment by chalov-anton** [comment]
:heavy_check_mark: Great job

The **Asset** filter works on: _Product_, _Page_, _Category_, and _Blocks_ pages.
Added a scenario to Regression Scope - https://studio.cucumber.io/projects/131313/test-plan/folders/1320712/scenarios/4951024

**Comment by ghost** [comment]
Hi @Nazar65, thank you for your contribution! </br>Please, complete [Contribution Survey](https://opensource.magento.com/survey/contribution/magento+adobe-stock-integration+1587), it will take less than a minute. </br>Your feedback will help us to improve contribution process.

**Pull request merged**